### PR TITLE
Latin hypercube sampler

### PR DESCRIPTION
**Pull request by aeslaughter**

### CODE REVIEW

**Comment by aeslaughter** [comment]
https://mooseframework.inl.gov/docs/PRs/14842/site/source/samplers/LatinHypercubeSampler.html

**Comment by aeslaughter** [comment]
I still need to do a bit of verification on the data produced.

File: modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+void
+LatinHypercubeSampler::sampleSetUp()
+{
+  // Setup bin information
+  const std::size_t num_dists = _distributions.size();
+  _size_bins.resize(num_dists);
+  _num_bins = _num_bins_input.size() == 1 ? std::vector<unsigned int>(_num_bins_input[0], num_dists)
+                                          : _num_bins_input;
+
+  for (std::size_t dist_idx = 0; dist_idx < num_dists; ++dist_idx)
```
Comment (jiangwen84) [inline modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C]: use range-based for loop here?

**Comment by aeslaughter** [comment]
OK, I fixed a little logic bug.

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Just a few minor comments right now.

File: modules/stochastic_tools/doc/content/source/samplers/LatinHypercubeSampler.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      layout={'xaxis':{'title':'Uniform', 'linewidth':1, 'showline':1, 'tickmode':'array', 'tickvals':[2004,2005,2006,2007,2008,2009,2010]},
+              'yaxis':{'title':'Normal', 'linewidth':1, 'showline':1, 'tickmode':'array', 'tickvals':[1970.7,1976.8,1978.3,1979.5,1980.5,1981.7,1983.2,1989.3]},
+              'width':775, 'height':800, 'font':{'size':16}, 'margin':{'l':100}}
+
+
+!syntax parameters /Samplers/LatinHypercube
+
+!syntax inputs /Samplers/LatinHypercube
+
+!syntax children /Samplers/LatinHypercube
```
[the remark this answers is not in the document]
Comment (aeslaughter) [reply modules/stochastic_tools/doc/content/source/samplers/LatinHypercubeSampler.md]: That is no longer needed, it is automatically added if there is a cite command.

File: modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C (hunk 1 above)
Comment (aeslaughter) [reply modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C]: No, the index is used to access two different lists.

File: modules/stochastic_tools/doc/content/source/samplers/LatinHypercubeSampler.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/stochastic_tools/doc/content/source/samplers/LatinHypercubeSampler.md]: The reference is not shown in https://mooseframework.inl.gov/docs/PRs/14842/site/source/samplers/LatinHypercubeSampler.html

File: framework/src/samplers/Sampler.C [hunk 3]
```diff
@@ -266,6 +267,15 @@ Sampler::getRand(const unsigned int index)
   return _generator.rand(index);
 }
 
+uint32_t
+Sampler::getRandl(unsigned int index, uint32_t lower, uint32_t upper)
+{
+  mooseAssert(index < _generator.size(), "The seed number index does not exists.");
+  TIME_SECTION(_perf_get_randl);
+
+  return _generator.randl(index, lower, upper);
```
Comment (permcody) [inline framework/src/samplers/Sampler.C]: I bet this that this is a little too small of a scope to time. I would recommend that after you look at this a bit that you take it back out. The timing code has got to be on the same order of the random generation code.

File: modules/stochastic_tools/include/samplers/LatinHypercubeSampler.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  std::vector<Distribution const *> _distributions;
+
+  /// Distribution names
+  const std::vector<DistributionName> & _distribution_names;
+
+  /// Number of intervals for each distribution
+  const std::vector<unsigned int> & _num_bins_input;
+
+  /// Upper limit to probability for each distribution (use double b/c MooseRandom uses double)
+  const std::vector<double> & _upper_limits;
```
Comment (permcody) [inline modules/stochastic_tools/include/samplers/LatinHypercubeSampler.h]: Sampler inherits from MooseObject so it's tightly tied into MOOSE and MOOSE types. Is there any reason not to use `Real`?

File: modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+  if (_lower_limits.size() != _distributions.size())
+    paramError("lower_limits",
+               "The length of 'lower_limits' must match the length of 'distributions'");
+
+  if (_num_bins_input.size() != 1 && _num_bins_input.size() != _distributions.size())
+    paramError("num_bins",
+               "The length of 'num_bins' must be one or match the length of 'distributions'");
+
+  for (const auto & lim : _upper_limits)
```
Comment (permcody) [inline modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C]: I'm 90% sure that range checking works for vectors. You might try it!

File: modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  setNumberOfRandomSeeds(2); // 0 = Bin seed; 1 = distribution sample seed
+  setNumberOfRows(getParam<dof_id_type>("num_rows"));
+  setNumberOfCols(_distributions.size());
+}
+
+void
+LatinHypercubeSampler::sampleSetUp()
+{
+  // Setup bin information
+  const std::size_t num_dists = _distributions.size();
```
Comment (permcody) [inline modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C]: you really hate `auto`

**Review by permcody — DISMISSED** [review]
I'm approving - you can merge or address comments.

**Review by jiangwen84 — DISMISSED** [review]

File: framework/src/samplers/Sampler.C (hunk 3 above)
Comment (aeslaughter) [reply framework/src/samplers/Sampler.C]: I will take those out, I was trying to figure out some timing issue last week when I added these.

File: modules/stochastic_tools/include/samplers/LatinHypercubeSampler.h (hunk 4 above)
Comment (aeslaughter) [reply modules/stochastic_tools/include/samplers/LatinHypercubeSampler.h]: The values get sent directly to MooseRandom, which uses double. It seemed like I should keep it the same.

File: modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C (hunk 6 above)
Comment (aeslaughter) [reply modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C]: I do.

File: modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C (hunk 5 above)
Comment (aeslaughter) [reply modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C]: I will try.

Comment (aeslaughter) [reply modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C]: It does support it, but the error messages for my version are a bit better. If I didn't already have them programmed and tested I would use it. Something for next time.

**Comment by aeslaughter** [comment]
@permcody I am done with this unless my response to the "double" vs "Real" isn't valid.

**Comment by permcody** [comment]
The only issue with the `double` -> `Real` stuff is that bleeds into everything. So now when you use Sampler interfaces, those will be `double` instead of `Real` too so in the code that calls those methods, should they also use `double`? What about further downstream? 

It might be the case that I made an error by using `double` in the first place in MooseRandom. The only time we might defensibly use `double` instead of Real is when the object can truly run outside of MOOSE. `MooseRandom` seems like something that shouldn't run outside MOOSE 😄. Perhaps we should consider changing the types in `MooseRandom`? You might think about doing all this sooner rather than later before you start getting a nasty mix of types everywhere in all your downstream code. I'm not going to hold this up though - I'll leave the consistency issue up to you.

File: framework/src/samplers/Sampler.C [hunk 7]
```diff
@@ -67,9 +67,7 @@ Sampler::Sampler(const InputParameters & parameters)
     _perf_get_global_samples(registerTimedSection("getGlobalSamples", 1)),
     _perf_get_local_samples(registerTimedSection("getLocalSamples", 1)),
     _perf_get_next_local_row(registerTimedSection("getNextLocalRow", 1)),
-    _perf_advance_generator(registerTimedSection("advanceGenerators", 2)),
```
Comment (permcody) [inline framework/src/samplers/Sampler.C]: Oops - you took this line out, but didn't take out the declaration.

**Review by permcody — CHANGES_REQUESTED** [review]

File: framework/src/samplers/Sampler.C (hunk 7 above)
Comment (aeslaughter) [reply framework/src/samplers/Sampler.C]: No, just the comma

**Comment by aeslaughter** [comment]
@permcody I will update MooseRandom and Sampler to use Real in another PR.

File: modules/stochastic_tools/doc/content/source/samplers/LatinHypercubeSampler.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (aeslaughter) [reply modules/stochastic_tools/doc/content/source/samplers/LatinHypercubeSampler.md]: Dang, it did when I rendered it locally. I will fix that.

**Comment by aeslaughter** [comment]
@permcody I was fixing some other stuff so I just updated MooseRandom to use Real, might as well pile it on for this PR

**Review by permcody — APPROVED** [review]

**Pull request merged**